### DISCUSSION TITLE
Issue seeing multiple intercepted requests

### DISCUSSION THREAD

**Discussion by jorlamd** [mitmproxy/mitmproxy#6913] [General]
Hi all! Hopefully I am posting in the right place.

I am trying to write an add-on that will listen to requests, and modify their parameters based on various criteria, and replay several requests with different combinations of modified parameters. I would like to see the original request go through, and all the modified ones intercepted and visible in the mitmproxy ui. At the moment I only see one request intercepted at a time, once I play that one, another intercepted one shows up, and so on. Any suggestions welcome. I am reading up on the topic anyhow and will post a solution if I get to it first.

```py
class CloneInterceptAddon:
    def __init__(self):
        self.tracked_signatures = set()

    async def request(self, flow: http.HTTPFlow) -> None:
        if "test" in flow.request.headers.keys():
            return
        conn_data = extract_connection_data(flow_to_dict(flow))
        for candidate_request in create_alternate_paths(conn_data.get("request").get("full_url")):
            await self.process_flow(flow, candidate_request)

    async def process_flow(self, flow: http.HTTPFlow, candidate: string) -> None:
        new_flow = flow.copy()
        new_flow.request.path = candidate
        new_flow.request.headers["test"] = "dup"
        if "view" in ctx.master.addons:
            ctx.master.commands.call("view.flows.duplicate", [new_flow])
        ctx.master.commands.call("replay.client", [new_flow])
        logger.info(f"Intercepting {candidate}")
        new_flow.intercept()

# Register the addon
addons = [
    CloneInterceptAddon()
]

```

I can see from the event log that the multiple calls are being created and intercepted, even if only one is shown at a time.

```
info: [20:59:42.945] Intercepting /r/MadeMeSmile/comments/1cpxacw/the_barcelona_superblocks_really_are_amazing_they/?chainedPosts=LoKol
info: [20:59:43.457] Intercepting /svc/shreddit/comments/r/mademesmile/t3_1cpxacw?render-mode=FmaCA&is_lit_ssr=false
info: [20:59:43.458] Intercepting /svc/shreddit/comments/r/mademesmile/t3_1cpxacw?render-mode=partial&is_lit_ssr=DIAHN
info: [20:59:43.461] Intercepting /svc/shreddit/pdp-right-rail/related/mademesmile/t3_1cpxacw?render-mode=tsHAI
info: [20:59:43.521] Intercepting /svc/shreddit/update-recaptcha?k=BjqUb
info: [20:59:43.525] Intercepting
...
```

https://github.com/mitmproxy/mitmproxy/assets/34373043/19f1a852-810d-4781-97fd-0d5c63cbc1b7

**Comment by Prinzhorn** [comment]
I have no idea what I'm looking at, could you please use curl and remove all the noise? It should be enough to make a single request with curl to show the issue you're having.

I assume your code is based on https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/duplicate-modify-replay.py? Does the example addon work as expected? If so, you should be able to work backwards to find the issue.

**Reply by jorlamd** [reply]
thanks for your reply! The add-on I am trying to build is not quite what you linked. The example you linked makes a modification, and straight replays the modified request. I'd like to, instead, be presented with multiple modified requests intercepted and ready for review and potential play in the UI.

In the video I shared you can see that I get one intercepted request, and once I play that one, I get the next. So, one at a time. Is that helpful? All I inteded to show in the recording is the single intercepted request highlighted in red.

Looking at the event log shared, I can see the calls to the routine that performs the interception are happening.

**Reply by jorlamd** [reply]
Much less cluttered demo of the actual behaviour, as requested. Thanks again.

https://github.com/mitmproxy/mitmproxy/assets/34373043/9a044b1a-37dc-4a11-81f5-7b192c6623ff

**Reply by Prinzhorn** [reply]
Can you provide a very minimal addon? I  don't know exactly how this is implemented, but my initial guess is that `flow.copy()` does not generate a new id. Could you add `new_flow.id = str(uuid.uuid4())` (with `import uuid`) right after `copy`?

**Reply by mhils** [reply]
It does generate a new id, https://github.com/mitmproxy/mitmproxy/blob/da05700b9a0b4dbacb763f3413be9abf50b6b20b/mitmproxy/coretypes/serializable.py#L56-L57 :)

**Reply by jorlamd** [reply]
While trying to provide you with a minimal example, I produced this. Which works! I am not sure why.

```py
from mitmproxy import http, ctx, command


class CloneInterceptAddon:

    async def request(self, flow: http.HTTPFlow) -> None:
        if flow.is_replay == "request":
            return
        await self.process_flow(flow)

    async def process_flow(self, flow: http.HTTPFlow) -> None:
        new_flow = flow.copy()
        flow.request.path = "/changed"
        if "view" in ctx.master.addons:
            ctx.master.commands.call("view.flows.duplicate", [new_flow])
        ctx.master.commands.call("replay.client", [new_flow])
        flow.intercept()

# Register the addon
addons = [
    CloneInterceptAddon()
]
```

**Reply by Prinzhorn** [reply]
The obvious difference is `flow.intercept()` vs `new_flow.intercept()`, right? You're saying this intercepts the copied flow?

**Reply by jorlamd** [reply]
No. The example I provided was incorrect.. Good catch. There were two bugs in that example. 

This is what I meant to share. This is a minimal implementation that has the original issue reported. For every request there is a modified one that gets intercepted, but I need to play one intercepted request to see the next.

```py
from mitmproxy import http, ctx, command


class CloneInterceptAddon:

    async def request(self, flow: http.HTTPFlow) -> None:
        if flow.is_replay == "request":
            return
        await self.process_flow(flow)

    async def process_flow(self, flow: http.HTTPFlow) -> None:
        new_flow = flow.copy()
        new_flow.request.path = "/changed"
        if "view" in ctx.master.addons:
            ctx.master.commands.call("view.flows.duplicate", [new_flow])
        ctx.master.commands.call("replay.client", [new_flow])
        new_flow.intercept()

# Register the addon
addons = [
    CloneInterceptAddon()
]
```
